src/main/java/org/broadinstitute/dsde/consent/ontology/datause/services/TextTranslationServiceImpl.java

```diff
@@ -89,13 +98,70 @@ public String translateDataset(String dataUseString) throws IOException {
         return translate(dataUse, TranslateFor.DATASET);
     }
 
+    @Override
+    public HashMap<String, Recommendation> translateParagraph(String paragraph) throws Exception {
+        return paragraph(paragraph);
+    }
+
     @Override
     public String translatePurpose(String dataUseString) throws IOException {
         ObjectMapper mapper = new ObjectMapper();
         DataUse dataUse = mapper.readValue(dataUseString, DataUse.class);
         return translate(dataUse, TranslateFor.PURPOSE);
     }
 
+    private HashMap<String, Recommendation> paragraph(final String paragraph) throws Exception {
+      HashMap<String, Recommendation> recommendations = new HashMap<>();
+
+      List<TermItem> terms = loadTermsFromGoogleStorage();
+
+      if (terms.isEmpty()) {
+        throw new Exception("Unable to process search terms");
+      }
+
+      for (TermItem term : terms) {
+        final String title = term.getTitle();
+        final String category = term.getCategory();
+        final String url = term.getUrl();
+        final String[] keywords = term.getKeywords();
+
+        for (String keyword : keywords) {
+          final boolean foundMatch = searchForKeyword(keyword, paragraph);
+
+          if (foundMatch) {
+            Recommendation recommendation = new Recommendation(
+                title,
+                category
+            );
+            if (!recommendations.containsKey(url)) {
+              recommendations.put(url, recommendation);
+            }
+          }
+        }
+      }
+
+      return recommendations;
+    }
+
+    private static boolean searchForKeyword(final String keyword, final String targetText) {
+      return StringUtils.containsIgnoreCase(targetText, keyword);
+    }
+
+    private List<TermItem> loadTermsFromGoogleStorage() {
+      try {
+        HttpResponse response = gcsStore.getStorageDocument("ontology/search-terms.json");
+        String terms = response.parseAsString();
+        return new Gson().fromJson(
+            terms,
+            new TypeToken<List<TermItem>>() {
+            }.getType()
+        );
+      } catch (Exception e) {
+        log.error("Error processing search terms from GCS: " + e.getMessage());
```

Comment:
Logging without an exception instead of throwing causes the exception's stack crawl to be lost. In this case I would rethrow using the caught exception, since you're going to throw anyway in the caller. That way the caller doesn't need to check for `isEmpty()` anymore.
```suggestion
        throw new Exception("Error processing search terms from GCS: ", e);
```

If you really don't want to rethrow, I would at least log the exception itself so the stack crawl for the error shows up in the logs.
```suggestion
        log.error("Error processing search terms from GCS: ", e);
```

Reply:
I like option 2, thanks!